File: sdk/python/examples/test_custom_skills_agents.py
#!/usr/bin/env python3
"""
A3S Code Python SDK - Custom Skills and Agents Integration Test

Tests loading and execution of custom skills and agents from ~/.a3s directory.

Run with: python examples/test_custom_skills_agents.py
"""

import asyncio
from pathlib import Path
from a3s_code import Agent


class CustomSkillsAgentsTest:
    """Integration test for custom skills and agents loading and execution."""

    def __init__(self, agent: Agent, config_path: str) -> None:
        self.agent = agent
        self.config_path = config_path
        self.skills_dir = self._find_skills_dir()
        self.agents_dir = self._find_agents_dir()

    @staticmethod
    def find_config_path() -> str:
        """Find config file in home directory or project root."""
        home_config = Path.home() / ".a3s" / "config.hcl"
        if home_config.exists():
            return str(home_config)

        # Try project root (6 levels up from examples/test_custom_skills_agents.py)
        project_config = Path(__file__).parent.parent.parent.parent.parent.parent / ".a3s" / "config.hcl"
        if project_config.exists():
            return str(project_config)

        raise FileNotFoundError("Config file not found. Please create ~/.a3s/config.hcl")

    @staticmethod
    def _find_skills_dir() -> str:
        """Find skills directory."""
        home_skills = Path.home() / ".a3s" / "skills"
        if home_skills.exists():
            return str(home_skills)

        # Try project root
        project_skills = Path(__file__).parent.parent.parent.parent.parent.parent / ".a3s" / "skills"
        if project_skills.exists():
            return str(project_skills)

        raise FileNotFoundError("Skills directory not found")

    @staticmethod
    def _find_agents_dir() -> str:
        """Find agents directory."""
        home_agents = Path.home() / ".a3s" / "agents"
        if home_agents.exists():
            return str(home_agents)

        # Try project root
        project_agents = Path(__file__).parent.parent.parent.parent.parent.parent / ".a3s" / "agents"
        if project_agents.exists():
            return str(project_agents)

        raise FileNotFoundError("Agents directory not found")

    @staticmethod
    def truncate(text: str, max_len: int) -> str:
        """Truncate text to max length."""
        if len(text) <= max_len:
            return text
        return f"{text[:max_len]}... (truncated)"

    async def test_custom_skills(self) -> None:
        """Test 1: Load custom skills in session."""
        print("\n📦 Test 1: Load Custom Skills in Session")
        print("-" * 80)

        session = self.agent.session(".", skill_dirs=[self.skills_dir])

        print("Testing: Ask about Rust best practices (should use rust-expert skill)...")
        result = session.send("What are the key principles of Rust ownership?")
        print(f"✓ Result preview: {self.truncate(result.text, 200)}")
        print()

        print("Testing: Ask about API design (should use api-designer skill)...")
        result = session.send("What are REST API best practices for error handling?")
        print(f"✓ Result preview: {self.truncate(result.text, 200)}")
        print()

        print("✅ Test 1 passed: Custom skills loaded and used in session\n")

    async def test_custom_agents(self) -> None:
        """Test 2: Load custom agents in session."""
        print("🤖 Test 2: Load Custom Agents in Session")
        print("-" * 80)

        session = self.agent.session(".", agent_dirs=[self.agents_dir])

        print("Testing: Request code review (should use code-reviewer agent)...")
        result = session.send("Review this function for potential issues: def add(a, b): return a + b")
        print(f"✓ Result preview: {self.truncate(result.text, 200)}")
        print()

        print("Testing: Request documentation (should use documentation-writer agent)...")
        result = session.send("Write API documentation for a user registration endpoint")
        print(f"✓ Result preview: {self.truncate(result.text, 200)}")
        print()

        print("✅ Test 2 passed: Custom agents loaded and used in session\n")

    async def test_skills_and_agents_together(self) -> None:
        """Test 3: Load both skills and agents together."""
        print("🔧 Test 3: Load Both Skills and Agents Together")
        print("-" * 80)

        session = self.agent.session(".", skill_dirs=[self.skills_dir], agent_dirs=[self.agents_dir])

        print("Testing: Complex task using both skills and agents...")
        result = session.send("Generate unit tests for a Python function that parses JSON")
        print(f"✓ Result preview: {self.truncate(result.text, 200)}")
        print()

        print("✅ Test 3 passed: Both skills and agents work together\n")

    async def test_multiple_sessions(self) -> None:
        """Test 4: Multiple sessions with different configurations."""
        print("🔀 Test 4: Multiple Sessions with Different Configurations")
        print("-" * 80)

        # Session 1: Only skills
        session1 = self.agent.session(".", skill_dirs=[self.skills_dir])

        # Session 2: Only agents
        session2 = self.agent.session(".", agent_dirs=[self.agents_dir])

        print("Testing: Session 1 with skills only...")
        result1 = session1.send("Explain Python decorators")
        print(f"✓ Session 1 result: {self.truncate(result1.text, 150)}")
        print()

        print("Testing: Session 2 with agents only...")
        result2 = session2.send("Review this code for improvements")
        print(f"✓ Session 2 result: {self.truncate(result2.text, 150)}")
        print()

        print("✅ Test 4 passed: Multiple sessions with different configs work correctly\n")

    async def run_all(self) -> None:
        """Run all custom skills and agents tests."""
        print("🚀 A3S Code Python SDK - Custom Skills and Agents Integration Test\n")
        print("=" * 80)

        print(f"📄 Using config: {self.config_path}")
        print(f"📚 Skills directory: {self.skills_dir}")
        print(f"🤖 Agents directory: {self.agents_dir}")

        print("=" * 80)

        await self.test_custom_skills()
        await self.test_custom_agents()
        await self.test_skills_and_agents_together()
        await self.test_multiple_sessions()

        print()
        print("=" * 80)
        print("✅ All custom skills and agents tests completed successfully!")
        print("=" * 80)


if __name__ == "__main__":
    config_path = CustomSkillsAgentsTest.find_config_path()
    agent = Agent.create(config_path)
    test = CustomSkillsAgentsTest(agent, config_path)
    asyncio.run(test.run_all())

File: sdk/node/examples/test_custom_skills_agents.ts
#!/usr/bin/env npx tsx
/**
 * A3S Code Node.js SDK - Custom Skills and Agents Integration Test (TypeScript)
 *
 * Tests loading and execution of custom skills and agents from ~/.a3s directory.
 *
 * Run with: npx tsx examples/test_custom_skills_agents.ts
 */

import { Agent, Session, AgentResult } from '../index.js';
import * as fs from 'fs';
import * as path from 'path';
import * as os from 'os';

class CustomSkillsAgentsTest {
  private readonly agent: Agent;
  private readonly configPath: string;

  constructor(agent: Agent, configPath: string) {
    this.agent = agent;
    this.configPath = configPath;
  }

  // ── Static Helpers ───────────────────────────────────────────────────────

  /**
   * Find config file in home directory or project root.
   */
  static findConfigPath(): string {
    const homeConfig: string = path.join(os.homedir(), '.a3s', 'config.hcl');
    if (fs.existsSync(homeConfig)) {
      return homeConfig;
    }

    // Try project root (6 levels up from examples/test_custom_skills_agents.ts)
    const projectConfig: string = path.join(__dirname, '..', '..', '..', '..', '..', '..', '.a3s', 'config.hcl');
    if (fs.existsSync(projectConfig)) {
      return projectConfig;
    }

    throw new Error('Config file not found. Please create ~/.a3s/config.hcl');
  }

  /**
   * Find skills directory.
   */
  static findSkillsDir(): string {
    const homeSkills: string = path.join(os.homedir(), '.a3s', 'skills');
    if (fs.existsSync(homeSkills)) {
      return homeSkills;
    }

    // Try project root
    const projectSkills: string = path.join(__dirname, '..', '..', '..', '..', '..', '..', '.a3s', 'skills');
    if (fs.existsSync(projectSkills)) {
      return projectSkills;
    }

    throw new Error('Skills directory not found');
  }

  /**
   * Find agents directory.
   */
  static findAgentsDir(): string {
    const homeAgents: string = path.join(os.homedir(), '.a3s', 'agents');
    if (fs.existsSync(homeAgents)) {
      return homeAgents;
    }

    // Try project root
    const projectAgents: string = path.join(__dirname, '..', '..', '..', '..', '..', '..', '.a3s', 'agents');
    if (fs.existsSync(projectAgents)) {
      return projectAgents;
    }

    throw new Error('Agents directory not found');
  }

  /**
   * Truncate text to max length.
   */
  static truncate(text: string, maxLen: number): string {
    if (text.length <= maxLen) {
      return text;
    }
    return `${text.substring(0, maxLen)}... (truncated)`;
  }

  // ── Instance Test Methods ────────────────────────────────────────────────

  /**
   * Test 1: Load custom skills in session.
   */
  async testCustomSkills(): Promise<void> {
    console.log('\n📦 Test 1: Load Custom Skills in Session');
    console.log('-'.repeat(80));

    const skillsDir: string = CustomSkillsAgentsTest.findSkillsDir();

    const session: Session = this.agent.session('.', { skillDirs: [skillsDir] });

    console.log('Testing: Ask about Rust best practices (should use rust-expert skill)...');
    const result1: AgentResult = await session.send('What are the key principles of Rust ownership?');
    console.log(`✓ Result preview: ${CustomSkillsAgentsTest.truncate(result1.text, 200)}`);
    console.log();

    console.log('Testing: Ask about API design (should use api-designer skill)...');
    const result2: AgentResult = await session.send('What are REST API best practices for error handling?');
    console.log(`✓ Result preview: ${CustomSkillsAgentsTest.truncate(result2.text, 200)}`);
    console.log();

    console.log('✅ Test 1 passed: Custom skills loaded and used in session\n');
  }

  /**
   * Test 2: Load custom agents in session.
   */
  async testCustomAgents(): Promise<void> {
    console.log('🤖 Test 2: Load Custom Agents in Session');
    console.log('-'.repeat(80));

    const agentsDir: string = CustomSkillsAgentsTest.findAgentsDir();

    const session: Session = this.agent.session('.', { agentDirs: [agentsDir] });

    console.log('Testing: Request code review (should use code-reviewer agent)...');
    const result1: AgentResult = await session.send('Review this function for potential issues: function add(a, b) { return a + b; }');
    console.log(`✓ Result preview: ${CustomSkillsAgentsTest.truncate(result1.text, 200)}`);
    console.log();

    console.log('Testing: Request documentation (should use documentation-writer agent)...');
    const result2: AgentResult = await session.send('Write API documentation for a user registration endpoint');
    console.log(`✓ Result preview: ${CustomSkillsAgentsTest.truncate(result2.text, 200)}`);
    console.log();

    console.log('✅ Test 2 passed: Custom agents loaded and used in session\n');
  }

  /**
   * Test 3: Load both skills and agents together.
   */
  async testSkillsAndAgentsTogether(): Promise<void> {
    console.log('🔧 Test 3: Load Both Skills and Agents Together');
    console.log('-'.repeat(80));

    const skillsDir: string = CustomSkillsAgentsTest.findSkillsDir();
    const agentsDir: string = CustomSkillsAgentsTest.findAgentsDir();

    const session: Session = this.agent.session('.', {
      skillDirs: [skillsDir],
      agentDirs: [agentsDir]
    });

    console.log('Testing: Complex task using both skills and agents...');
    const result: AgentResult = await session.send('Generate unit tests for a JavaScript function that parses JSON');
    console.log(`✓ Result preview: ${CustomSkillsAgentsTest.truncate(result.text, 200)}`);
    console.log();

    console.log('✅ Test 3 passed: Both skills and agents work together\n');
  }

  /**
   * Test 4: Multiple sessions with different configurations.
   */
  async testMultipleSessions(): Promise<void> {
    console.log('🔀 Test 4: Multiple Sessions with Different Configurations');
    console.log('-'.repeat(80));

    const skillsDir: string = CustomSkillsAgentsTest.findSkillsDir();
    const agentsDir: string = CustomSkillsAgentsTest.findAgentsDir();

    // Session 1: Only skills
    const session1: Session = this.agent.session('.', { skillDirs: [skillsDir] });

    // Session 2: Only agents
    const session2: Session = this.agent.session('.', { agentDirs: [agentsDir] });

    console.log('Testing: Session 1 with skills only...');
    const result1: AgentResult = await session1.send('Explain JavaScript async/await');
    console.log(`✓ Session 1 result: ${CustomSkillsAgentsTest.truncate(result1.text, 150)}`);
    console.log();

    console.log('Testing: Session 2 with agents only...');
    const result2: AgentResult = await session2.send('Review this code for improvements');
    console.log(`✓ Session 2 result: ${CustomSkillsAgentsTest.truncate(result2.text, 150)}`);
    console.log();

    console.log('✅ Test 4 passed: Multiple sessions with different configs work correctly\n');
  }

  // ── Run All ───────────────────────────────────────────────────────────────

  async runAll(): Promise<void> {
    console.log('🚀 A3S Code Node.js SDK - Custom Skills and Agents Integration Test\n');
    console.log('='.repeat(80));

    console.log(`📄 Using config: ${this.configPath}`);

    const skillsDir: string = CustomSkillsAgentsTest.findSkillsDir();
    console.log(`📚 Skills directory: ${skillsDir}`);

    const agentsDir: string = CustomSkillsAgentsTest.findAgentsDir();
    console.log(`🤖 Agents directory: ${agentsDir}`);

    console.log('='.repeat(80));

    await this.testCustomSkills();
    await this.testCustomAgents();
    await this.testSkillsAndAgentsTogether();
    await this.testMultipleSessions();

    console.log();
    console.log('='.repeat(80));
    console.log('✅ All custom skills and agents tests completed successfully!');
    console.log('='.repeat(80));
  }
}

async function main(): Promise<void> {
  const configPath = CustomSkillsAgentsTest.findConfigPath();
  const agent = await Agent.create(configPath);
  const test = new CustomSkillsAgentsTest(agent, configPath);
  await test.runAll();
}

// Run tests
main().catch((err: unknown) => {
  console.error('❌ Test failed:', err);
  process.exit(1);
});
